causallearn/search/FCMBased/PNL/PNL.py

```diff
@@ -115,56 +100,54 @@ def nica_mnd(self, X, TotalEpoch, KofMix):
         Y (n*T): the separation result.
         """
         trpattern = X.T
-        trpattern = trpattern - np.tile(np.mean(trpattern, axis=1).reshape(2, 1), (1, len(trpattern[0])))
-        trpattern = np.dot(np.diag(1.5 / np.std(trpattern, axis=1)), trpattern)
+
         # --------------------------------------------------------
         x1 = torch.from_numpy(trpattern[0, :]).type(torch.FloatTensor).reshape(-1, 1)
         x2 = torch.from_numpy(trpattern[1, :]).type(torch.FloatTensor).reshape(-1, 1)
         x1.requires_grad = True
         x2.requires_grad = True
+
         y1 = x1
 
         Final_y2 = x2
         Min_loss = float('inf')
 
-        G1 = MLP(1, 1, n_layers=1, n_units=20)
-        G2 = MLP(1, 1, n_layers=1, n_units=20)
-        # MixGaussian = MixGaussianLayer(Mix_K=KofMix)
-        G3 = MLP(1, 1, n_layers=1, n_units=20)
+        G1 = MLP(1, 1, n_layers=3, n_units=12)
+        G2 = MLP(1, 1, n_layers=1, n_units=12)
         optimizer = torch.optim.Adam([
             {'params': G1.parameters()},
-            {'params': G2.parameters()},
-            {'params': G3.parameters()}], lr=1e-4, betas=(0.9, 0.99))
+            {'params': G2.parameters()}], lr=1e-5, betas=(0.9, 0.99))
+
+        loss_all = torch.zeros(TotalEpoch)
+        loss_pdf_all = torch.zeros(TotalEpoch)
+        loss_jacob_all = torch.zeros(TotalEpoch)
 
         for epoch in range(TotalEpoch):
+            G1.zero_grad()
+            G2.zero_grad()
```

Comment:
will this change the original logic?

Reply:
`zero_grad()` should be added to clear the gradients. This is an error of the previous implementation.